### PR TITLE
Adds support for `useBackgroundQuery` and `useReadQuery` in Streaming SSR

### PR DESCRIPTION
**Pull request by alessbell**
This PR creates a new class, `NextSSRApolloClient` in `package/src/ssr/NextSSRApolloClient.tsx`, which can be imported and used in **client components** where one would use `ApolloClient` outside of the app directory.

This extension of the `ApolloClient` class registers "background queries" (queries triggered by components rendering on the server) via the same `registerLateInitializingQueue` strategy already used to register results on the client. By keeping track of queries already run on the server, we can avoid duplicating these requests on the client when the same components are executed in the browser, since the result will be streamed in once it has completed.

### CODE REVIEW

File: examples/hack-the-supergraph-ssr/app/error.tsx [hunk 1]
```diff
@@ -4,16 +4,14 @@ import React from "react";
...
-  children,
-  code,
   error,
-}: React.PropsWithChildren<{
-  code?: string;
-  error: { toString(): string };
-}>) => (
+  reset,
+}: {
+  error: Error;
```
Comment (alessbell) [inline examples/hack-the-supergraph-ssr/app/error.tsx:10]: Updating our error pages to conform with the API as described in the latest docs: https://nextjs.org/docs/app/api-reference/file-conventions/error

Comment (alessbell) [reply examples/hack-the-supergraph-ssr/app/error.tsx:10]: Interestingly, there's a lint error if I try to type `error` as `ApolloError` here:

CleanShot 2023-06-14 at 11 46 38

File: examples/polls-demo/app/cc/loading.tsx [hunk 2]
```diff
@@ -1,5 +0,0 @@
-import { PollSkeleton } from "@/components/poll";
```
Comment (alessbell) [inline examples/polls-demo/app/cc/loading.tsx:1]: I'll add this back - I temporarily removed it for testing with `useBackgroundQuery` where using a `Suspense` boundary directly was preferred

File: package/.eslintrc.js [hunk 3]
```diff
@@ -20,5 +20,11 @@ module.exports = {
   rules: {
     "@typescript-eslint/no-explicit-any": "off",
     "react/prop-types": "off",
+    "@typescript-eslint/no-unused-vars": [
+      "warn",
+      {
+        varsIgnorePattern: "^_",
```
Comment (alessbell) [inline package/.eslintrc.js:26]: Added this for array destructuring where the first item isn't needed

File: package/src/ssr/ApolloRehydrateSymbols.tsx [hunk 4]
```diff
@@ -1,14 +1,23 @@
...
+  BackgroundQueriesCache,
+} from "./types";
 import type { DataTransport } from "./dataTransport";
 
 declare global {
   interface Window {
     [ApolloRehydrationCache]?: RehydrationCache;
     [ApolloResultCache]?: ResultsCache;
     [ApolloSSRDataTransport]?: DataTransport<SuperJSONResult>;
+    // TODO: types
```
Comment (alessbell) [inline package/src/ssr/ApolloRehydrateSymbols.tsx:14]: I think I already addressed this, will remove the comment

File: package/src/ssr/RehydrationContext.tsx [hunk 5]
```diff
@@ -13,20 +14,18 @@ const ApolloRehydrationContext = React.createContext<
...
-    if (cache instanceof NextSSRInMemoryCache) {
-      cache.setRehydrationContext(rehydrationContext.current);
-    } else {
-      throw new Error(
-        "When using Next SSR, you must use the `NextSSRInMemoryCache`"
-      );
-    }
+    if (client instanceof NextSSRApolloClient) {
+      client.setRehydrationContext(rehydrationContext.current);
+    } // else error?
```
Comment (alessbell) [inline package/src/ssr/RehydrationContext.tsx:25]: Oops missed this comment, adding these errors shortly..

File: package/README.md [hunk 6]
```diff
@@ -164,7 +164,12 @@ export default function RootLayout({
 > ☝️ This will work even if your layout is a React Server Component and will also allow the children of the layout to be React Server Components.  
 > It just makes sure that all Client Components will have access to the same Apollo Client instance, shared through the `ApolloNextAppProvider`.
 
-Now you can use the hooks `useQuery`, `useSuspenseQuery`, `useFragment`, and `useApolloClient` from `"@apollo/experimental-nextjs-app-support/ssr"` in your Client components like you are used to.
+You can import the following Apollo Client hooks from `"@apollo/experimental-nextjs-app-support/ssr"` in your client components (make sure you are not importing these hooks from `@apollo/client` as this package wraps and re-exports them to support streaming SSR):
```
Comment (alessbell) [inline package/README.md:167]: Just made these hooks a list to make this section a bit harder to miss.

I removed `useApolloClient` from this list since that seemed no longer relevant, but let me know if that should be added back @phryneas 👍

File: package/src/ssr/NextSSRApolloClient.tsx [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  > & { uninitialized?: boolean } = {
+    incomingBackgroundQueries: [],
+    uninitialized: true,
+  };
+
+  constructor(options: ApolloClientOptions<TCacheShape>) {
+    super({
+      ...options,
+      // TODO: the memoization in `documentTransform` can be removed once
+      // https://github.com/apollographql/apollo-client/pull/10968 is merged.
```
Comment (alessbell) [inline package/src/ssr/NextSSRApolloClient.tsx:37]: This was just merged, I should be able to remove this once the next AC beta is released 🎉

Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx:37]: After upgrading to beta 3, the client-side fetch is no longer being deduped. Investigating...

File: package/src/ssr/NextSSRApolloClient.tsx [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    const result = super.watchQuery(options);
+    return result;
+  }
+
+  setRehydrationContext(rehydrationContext: RehydrationContextValue) {
+    if (this.rehydrationContext.uninitialized) {
+      rehydrationContext.incomingBackgroundQueries.push(
+        ...this.rehydrationContext.incomingBackgroundQueries
+      );
+    }
```
Comment (phryneas) [inline package/src/ssr/NextSSRApolloClient.tsx:165]: Maybe here we could even move all data from the old `incombineBackgroundQueries` to the new ones?

Something like

```suggestion
    if (rehydrationContext.incomingBackgroundQueries !== this.rehydrationContext.incomingBackgroundQueries)
      rehydrationContext.incomingBackgroundQueries.push(
        ...this.rehydrationContext.incomingBackgroundQueries.splice(0, Infinity)
      )
```

Then we could skip all the `uninitialized` handling altogether.

File: package/src/ssr/NextSSRApolloClient.tsx [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+                this["queryManager"].fetchCancelFns.delete(cacheKey);
+
+              const [_, reject] = this.resolveFakeQueries.get(cacheKey) ?? [];
+
+              this["queryManager"].fetchCancelFns.set(
+                cacheKey,
+                (reason: unknown) => {
+                  cleanupCancelFn();
+                  if (reject) {
+                    this.resolveFakeQueries.delete(cacheKey);
```
Comment (phryneas) [inline package/src/ssr/NextSSRApolloClient.tsx:123]: Maybe instead of calling this in multiple places, we could have a `promise.finally(...)` call with this directly after the definition of `promise`?

File: package/src/ssr/NextSSRApolloClient.tsx [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+                    .catch((err) => {
+                      observer.error(err);
+                    });
+                })
+              );
+
+              const cleanupCancelFn = () =>
+                this["queryManager"].fetchCancelFns.delete(cacheKey);
+
+              const [_, reject] = this.resolveFakeQueries.get(cacheKey) ?? [];
```
Comment (phryneas) [inline package/src/ssr/NextSSRApolloClient.tsx:116]: This line should probably better go into the `fetchCancelFn` definition a few lines down - otherwise `reject` would never be undefined, and also execute if the query had resolved successfully.

File: package/src/ssr/NextSSRInMemoryCache.tsx [hunk 11]
```diff
@@ -18,21 +16,6 @@ export class NextSSRInMemoryCache extends InMemoryCache {
...
-      if (Array.isArray(window[ApolloResultCache] || [])) {
-        registerLateInitializingQueue(ApolloResultCache, (data) =>
-          this.write(data)
-        );
-      } else {
-        throw new Error(
-          "On the client side, only one instance of `NextSSRInMemoryCache` can be created!"
-        );
-      }
-    }
```
[the remark this answers is not in the document]
Comment (phryneas) [reply package/src/ssr/NextSSRInMemoryCache.tsx:35]: Huh, good question. If a real `watchQuery/fetchQuery` call latches onto our fake query before it finishes, it wouldn't be necessary - but if it doesn't, it would be necessary.
So maybe we could create the fake ongoing query, and then immediately call `QueryManager.fetchQuery` to make sure its' results gets written into the store? Or we just call `write` after resolving the query? (The latter might have a lower risk of race conditions 🤔 )

**Review by phryneas — CHANGES_REQUESTED** [review]
Already some feedback. I'm curious why that deduplication change didn't do what we expected, though. Gonna look a little deeper 👀

**Comment by phryneas** [comment]
@alessbell I found out what was going wrong.

We need a `getQuerymanager(this)` method instead of just accessing `this['QueryManager']`, as that one is `any`, not `QueryManager<TCacheShape>`. Add that and you'll immediately notice ;)

File: package/src/ssr/NextSSRApolloClient.tsx (hunk 7 above)
Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx]: Fixed in [9a74dc6](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/9a74dc607f20aa1239cfdeee31f0e0eb157b5e43)

File: package/src/ssr/NextSSRInMemoryCache.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (alessbell) [reply package/src/ssr/NextSSRInMemoryCache.tsx:35]: Sounds good :) Addressed in [9a74dc6](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/9a74dc607f20aa1239cfdeee31f0e0eb157b5e43).

File: package/src/ssr/NextSSRApolloClient.tsx (hunk 10 above)
Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx]: Fixed in [c49d481](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/c49d481ff5c9bdd474c4b61ba65b610162ffe067).

File: package/src/ssr/NextSSRApolloClient.tsx (hunk 8 above)
Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx]: Fixed in [c49d481](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/c49d481ff5c9bdd474c4b61ba65b610162ffe067)

File: package/src/ssr/NextSSRApolloClient.tsx (hunk 9 above)
Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx]: Addressed in [fa82b94](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/fa82b94c0189747b9463628c1c57b8d74803d1af)

**Comment by alessbell** [comment]
Thanks for the review, @phryneas! I believe I've gotten to all the feedback (including offline comment re: test helpers addressed in [037c8ba](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/037c8ba783feaf46f6e81d9fb3259142e9cdb67e)) and I've done some more manual testing. This is ready for a final look :)

File: examples/polls-demo/app/cc/loading.tsx (hunk 2 above)
Comment (phryneas) [reply examples/polls-demo/app/cc/loading.tsx:1]: Reminder to add this back :)

File: package/src/ssr/hooks.ts [hunk 12]
```diff
@@ -21,6 +24,13 @@ export const useSuspenseQuery = wrap(orig_useSuspenseQuery, [
   "data",
   "networkStatus",
 ]);
+export const useReadQuery = wrap(orig_useReadQuery, ["data", "networkStatus"]);
+// @ts-ignore
+export const useBackgroundQuery: typeof orig_useBackgroundQuery = (...args) => {
+  useRehydrationContext();
+  // @ts-ignore
+  return orig_useBackgroundQuery(...args);
+};
```
Comment (phryneas) [inline package/src/ssr/hooks.ts:33]: let's get rid of those `@ts-ignore` annotations so eslint shuts up :)
```suggestion
export const useBackgroundQuery: typeof orig_useBackgroundQuery = (...args: [any, any]) => {
  useRehydrationContext();
  return orig_useBackgroundQuery(...args);
};
```

File: package/src/ssr/NextSSRApolloClient.tsx [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+                  promise
+                    .then((result) => {
+                      observer.next(result);
+                      observer.complete();
+                    })
+                    .catch((err) => {
+                      observer.error(err);
+                    }).finally(() => {
+                      this.resolveFakeQueries.delete(cacheKey);
+                    });
```
Comment (phryneas) [inline package/src/ssr/NextSSRApolloClient.tsx:105]:
```suggestion
                    }).finally(() => {
                      this.resolveFakeQueries.delete(cacheKey);
                      cleanupCancelFn();
                    });
```

For comparison:
https://github.com/apollographql/apollo-client/blob/d64a9656ee745349bb4958a743065f2a91c2d0cc/src/core/QueryManager.ts#L1278

File: package/src/ssr/NextSSRApolloClient.tsx [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+                })
+              );
+              const queryManager = getQueryManager<TCacheShape>(this);
```
Comment (phryneas) [inline package/src/ssr/NextSSRApolloClient.tsx:108]:
```suggestion
```

This should already be in scope due to line 77.

**Review by phryneas — CHANGES_REQUESTED** [review]
This looks very good! 
Only some minor little changes, then this should be good to go!

File: package/src/ssr/hooks.ts (hunk 12 above)
Comment (alessbell) [reply package/src/ssr/hooks.ts]: Resolved in [867d581](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/867d5812951b586c6c8cc23fba07e82e0dd129bf)

File: package/src/ssr/NextSSRApolloClient.tsx (hunk 14 above)
Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx]: Ah yep, thanks for catching that! Resolved via [6462b2a](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/6462b2a8c2f1dfb24a66a34f77944cb1a9436546)

File: package/src/ssr/NextSSRApolloClient.tsx (hunk 13 above)
Comment (alessbell) [reply package/src/ssr/NextSSRApolloClient.tsx]: Resolved in [6462b2a](https://github.com/apollographql/apollo-client-nextjs/pull/38/commits/6462b2a8c2f1dfb24a66a34f77944cb1a9436546)
